### PR TITLE
Implement range expression

### PR DESCRIPTION
**Pull request by sbezverk**
Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>
Closes: #19

### CODE REVIEW

**Comment by sbezverk** [comment]
@stapelberg Please take a look, by some reason I get mismatch between want and got, appreciate some feedback.

File: expr/range.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	if e.Register > 0 {
+		attrs = append(attrs, netlink.Attribute{Type: unix.NFTA_RANGE_SREG, Data: binaryutil.BigEndian.PutUint32(e.Register)})
+	}
+	attrs = append(attrs, netlink.Attribute{Type: unix.NFTA_RANGE_OP, Data: binaryutil.BigEndian.PutUint32(uint32(e.Op))})
+	// |00012|N-|00003|	|len |flags| type|
+	// |00006|--|00001|	|len |flags| type|
+	// | 07 e8 00 00  |	|      data      |
+	if e.FromData > 0 {
+		attrs = append(attrs, netlink.Attribute{Length: 12, Type: unix.NLA_F_NESTED | unix.NFTA_RANGE_FROM_DATA, Data: []byte{}})
+		attrs = append(attrs, netlink.Attribute{Length: 6, Type: 0x1, Data: binaryutil.BigEndian.PutUint32(e.FromData)})
```
Comment (stapelberg) [inline expr/range.go]: Here and below, use unix.NFTA_DATA_VALUE instead of 0x1

File: expr/range.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	var attrs []netlink.Attribute
+	if e.Register > 0 {
+		attrs = append(attrs, netlink.Attribute{Type: unix.NFTA_RANGE_SREG, Data: binaryutil.BigEndian.PutUint32(e.Register)})
+	}
+	attrs = append(attrs, netlink.Attribute{Type: unix.NFTA_RANGE_OP, Data: binaryutil.BigEndian.PutUint32(uint32(e.Op))})
+	// |00012|N-|00003|	|len |flags| type|
+	// |00006|--|00001|	|len |flags| type|
+	// | 07 e8 00 00  |	|      data      |
+	if e.FromData > 0 {
+		attrs = append(attrs, netlink.Attribute{Length: 12, Type: unix.NLA_F_NESTED | unix.NFTA_RANGE_FROM_DATA, Data: []byte{}})
```
Comment (stapelberg) [inline expr/range.go]: Here and below, can you remove the Length attribute? It should be populated automatically.

File: expr/range.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	ToData   uint32
+}
+
+func (e *Range) marshal() ([]byte, error) {
+	var attrs []netlink.Attribute
+	if e.Register > 0 {
+		attrs = append(attrs, netlink.Attribute{Type: unix.NFTA_RANGE_SREG, Data: binaryutil.BigEndian.PutUint32(e.Register)})
+	}
+	attrs = append(attrs, netlink.Attribute{Type: unix.NFTA_RANGE_OP, Data: binaryutil.BigEndian.PutUint32(uint32(e.Op))})
+	// |00012|N-|00003|	|len |flags| type|
```
Comment (stapelberg) [inline expr/range.go]: Here and below, can you remove the comments please?

File: expr/range.go (hunk 1 above)
Comment (sbezverk) [reply expr/range.go]: Will do

File: expr/range.go (hunk 2 above)
Comment (sbezverk) [reply expr/range.go]: It is a problem, I spent half of day yesterday on this unit test, it was constantly failing, I was able to fix it only after specifying the length for both attributes.

File: expr/range.go (hunk 3 above)
Comment (sbezverk) [reply expr/range.go]: Will do.

**Comment by sbezverk** [comment]
/hold

Filed netlink issue: https://github.com/mdlayher/netlink/issues/140

**Comment by sbezverk** [comment]
@stapelberg Please check new revision, I think it is ready.

File: expr/range.go [hunk 4]
```diff
@@ -67,10 +77,23 @@ func (e *Range) unmarshal(data []byte) error {
...
 		}
 	}
 	return ad.Err()
 }
+
+func nestedAttr(data []byte, attrType uint16) ([]byte, error) {
+	ae := netlink.NewAttributeEncoder()
+	ae.Do(unix.NLA_F_NESTED|attrType, func() ([]byte, error) {
+		nae := netlink.NewAttributeEncoder()
+		// You can set endianness if you need to for the data.
```
Comment (stapelberg) [inline expr/range.go]: This comment looks like a left-over from debugging? Can you remove it?

File: nftables_test.go [hunk 5]
```diff
@@ -1620,8 +1620,191 @@ func TestConfigureRange(t *testing.T) {
...
+			},
+		},
+	})
+
+	if err := c.Flush(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestConfigureRangeIPv4s(t *testing.T) {
```
Comment (stapelberg) [inline nftables_test.go]: Here and below, why “IPv4s” instead of “IPv4”? What does the extra “s” stand for?

File: expr/range.go (hunk 4 above)
Comment (sbezverk) [reply expr/range.go]: done

File: nftables_test.go (hunk 5 above)
Comment (sbezverk) [reply nftables_test.go]: it was just indicating that there are multiple IPs, removed 's'

**Comment by sbezverk** [comment]
@stapelberg PTAL, I have addressed your comments.

**Comment by stapelberg** [comment]
Thanks!

**Pull request merged**